README.md
Webpage of the Lab of Andrea Pauli
==================================

This is the source of the webpage, the webpage itself is available at
[paulilab.org](http:/paulilab.org/).


Editing
-------

* [Editing files in the browser](https://help.github.com/articles/editing-files-in-your-repository/)
* [Markdown syntax](https://daringfireball.net/projects/markdown/syntax)


### Adding posts

Create an .md file under `/_posts` (start by copying an already existing file).
The filename must have the form `YYYY-MM-DD-some-post.md` to be valid.

A JPEG file with a post picture should be called `YYYYMMDD-SomeLabel.jpg` and
placed in `/assets/img/posts/`. Before adding the picture resize it to have the
larger side no bigger than 1000 px. This will make it load faster.


### Adding publications

Create an .md file under `/_pubs` (start by copying an already existing file).

If a whole publication year is missing, check that it has been added under
`years` inside the `/_config.yml` file.


### Adding/Removing people

Create an .md file under `/_people` (start by copying an already existing file).

A JPEG file with a person picture should be placed in `/assets/img/people/`.
Before adding the picture resize it to have a width of 108 px.


### splash image

The splash image is specified in the front matter using:

    splash: /assets/img/splash/…

If there is no `splash:` record, the `default_splash:` image from `_config.yml`
is used.

Splash image can also be specified for multiple pages at once using the
`splash:` entry in the corresponding layout file.


### Appearance in the search results and on Facebook/Twitter

For pages specify `description: "Some description..."` to have the text
displayed in the search results and on Facebook/Twitter.

For news posts description gets automatically generated from the post's
content.

Specifying `picture: /assets/img/some_picture.jpg` for any page or news post
results in the image being used on Facebook/Twitter.


### sitemap

A `sitemap.xml` is automatically build based on the post publish dates. If you
modify a post after it was published, add `last_modified_at` with the current
data to the front matter.

[Technical info on &lt;lastmod&gt; in
jekyll-sitemap](https://github.com/jekyll/jekyll-sitemap#lastmod-tag).


Deploying
---------

When editing in the browser, the website re-deploys automatically.

Otherwise, one needs to push to the repository using git:

    git push

After changing any javascript or less/css style files, the website has to be
rebuilt offline using grunt, see Developmental setup.


Developmental setup
-------------------

This is a setup for Mac OS. Linux shold be a lot simpler, see .travis.yml for
inspiration.


### Installs to compile js and less

1. Install nodejs from http://nodejs.org/en/download/ (last tested: 10.13.0)
   This also bundles npm.
2. Install the rest locally using package.json
    npm install
3. Install grunt-cli globally
    npm install -g grunt-cli
   or put `./node_modules/.bin/` on your `$PATH`.


### Installs to run jekyll and sass

1. Install recent ruby version from MacPorts (last tested: 2.5)
    sudo port install rb25-bundler
2. Make sure to put the bundle command on PATH (otherwise sass will fail)
    export PATH=/opt/local/libexec/ruby2.5/:$PATH
2. Install the rest locally
    bundle install --path vendor/bundle


### Developing

Running `grunt` (the default task) starts a local webserver on
http://0.0.0.0:3000/ with a live reload.


### Building and publishing the website

Any changes to layout or javascript code needs rebuilding the website using
`grunt`.  Running `grunt build` triggers a complete build of the website
regenerating the relevant CSS/JS assets (`/assets/css/main.min.css`,
`/assets/js/main.min.css` and `/assets/js/fun.min.css`).  The static website
(output of `jekyll`) is put into the `_site` subdirectory.  The changed assets
should be pushed to the github repository. The push will trigger a jekyll run
on the github side and publishing of the changes on the website.

Seeing changes after update might require a refresh in the browser, e.g., Cmd+R
in Firefox.

To ensure predictable results it is advisable to keep the [dependencies and
versions](https://pages.github.com/versions/) available to the jekyll run on
the github side in sync with the local setup (`Gemfile`).

#### Check the website speed

https://developers.google.com/speed/pagespeed/insights

### Favicon

The favicon is generated from the `logo.svg` template using the [real favicon
generator](http://realfavicongenerator.net/).


### SEO

Sitemaps get generated automatically by the jekyll-sitemap plugin.

https://www.bing.com/webmaster/home/dashboard

https://www.google.com/webmasters/tools

https://developers.facebook.com/tools/debug/og/object/

https://search.google.com/structured-data/testing-tool

https://www.seocentro.com/tools/seo/seo-analyzer.html

http://www.seowebpageanalyzer.com/


_pages/news.md
---
layout: default
title: Lab News
permalink: /news/
splash: /assets/img/splash/embryos.jpg
#description: "{{ site.posts[0].excerpt }}"
---

{% for post in site.posts %}
<div markdown="1">

#### {{ post.title }}

{{ post.date | date_to_string }}

{{ post.excerpt | remove: '<p>' | remove: '</p>' | replace: '\n', ' ' }}

[Read more &hellip;]({{ site.baseurl }}{{ post.url }})

</div>
{% endfor %}


tools/gen_maps_style.py
import os
import re
import numpy as np
from skimage.color import rgb2hsv, hsv2rgb
import matplotlib.colors as mplcolors


def make_blue(color):
    rgb = mplcolors.hex2color(color)
    rgb = np.array(rgb).reshape(1, 1, 3)
    hsv = rgb2hsv(rgb)
    print hsv
    #hsv[0, 0, 2] = 1 - hsv[0, 0, 2]
    hsv[0, 0, 0] = 145 / 255.
    hsv[0, 0, 1] = 40 / 100.
    rgb = hsv2rgb(hsv)
    rgb = rgb[0, 0]
    return mplcolors.rgb2hex(rgb)

if __name__ == '__main__':
    styles = {
            'blue': make_blue
            }
    name = 'blue'
    dirname = os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        '..',
        '_includes')
    color_re = '(#[0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]' + \
               '[0-9a-fA-F][0-9a-fA-F][0-9a-fA-F])'
    contents = open(os.path.join(dirname, 'maps_style_silver')).read()
    replacements = {}
    for match in re.finditer(color_re, contents):
        color = match.group(0)
        replacements[color] = styles[name](color)
    for src, dst in replacements.items():
        contents = contents.replace(src, dst)
    with open(os.path.join(dirname, 'maps_style_' + name), 'w') as fw:
        fw.write(contents)


Gruntfile.js
module.exports = function(grunt) {
  var log = function (err, stdout, stderr, cb) {
    console.log(stdout);
    cb();
  }
  require('load-grunt-tasks')(grunt, { scope: 'devDependencies' });
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    builddir: 'assets',
    bootstrapdir: 'node_modules/bootstrap/js/dist/',
    jquerydir: 'node_modules/jquery/dist/',
    popperjsdir: 'node_modules/popper.js/dist/',
    instafeeddir: 'node_modules/instafeed.js',
    banner: '/*!\n' +
    ' * <%= pkg.name %> v<%= pkg.version %>\n' +
    ' * Homepage: <%= pkg.homepage %>\n' +
    ' * Copyright 2017-<%= grunt.template.today("yyyy") %> <%= pkg.author %>\n' +
    ' * Licensed under <%= pkg.license %>\n' +
    '*/\n',
    sass: {
      dist: {
        options: {
          bundleExec: true,
          style: 'compressed',
          trace: true,
          sourcemap: 'none'
        },
        files: {}
      }
    },
    autoprefixer: {
      dist: {
        src: '*/main.css'
      }
    },
    concat: {
      options: {
        banner: '<%= banner %>',
        stripBanners: false
      },
      js: {
        src: [
          '<%=jquerydir%>/jquery.slim.js', // Bootstrap's dependency
          '<%=popperjsdir%>/umd/popper.js', // Bootstrap's dependency
          //'node_modules/bootstrap/dist/js/bootstrap.js',
          '<%=bootstrapdir%>/util.js', // Required by .navbar-collapse
          //'<%=bootstrapdir%>/alert.js',
          //'<%=bootstrapdir%>/button.js',
          //'<%=bootstrapdir%>/carousel.js',
          '<%=bootstrapdir%>/collapse.js', // Required by .navbar-collapse
          '<%=bootstrapdir%>/dropdown.js', // Required by .navbar-collapse
          //'<%=bootstrapdir%>/modal.js',
          //'<%=bootstrapdir%>/tooltip.js',
          //'<%=bootstrapdir%>/popover.js',
          //'<%=bootstrapdir%>/scrollspy.js',
          //'<%=bootstrapdir%>/tab.js',
          '<%=bootstrapdir%>/index.js',
        ],
        dest: '<%=builddir%>/js/main.js'
      },
      funjs: {
        src: [
          '<%= instafeeddir %>/instafeed.min.js',
          'js/fun.js'
        ],
        dest: '<%= builddir %>/js/fun.js'
      }
    },
    uglify: {
      options: {
        compress: {
          warnings: false
        },
        mangle: true,
        preserveComments: /^!|@preserve|@license|@cc_on/i
      },
      js: {
        src: '<%= concat.js.dest %>',
        dest: '<%= builddir %>/js/main.min.js'
      },
      funjs: {
        src: '<%= concat.funjs.dest %>',
        dest: '<%= builddir %>/js/fun.min.js'
      }
    },
    clean: {
      assets: [
        'assets/js/*.js', '!assets/js/*.min.js',
        'assets/css/*.css', '!assets/css/*.min.css'
      ]
    },
    uncss: {
      dist: {
        options: {
          ignore: [
            // needed (no idea why it gets uncss-ed)
            '.main img.post-picture',
            // needed for Bootstrap's pager buttons
            /(#|\.)pager.*/,
            // needed for Bootstrap's transitions
            '.bs.carousel',
            '.slid.bs.carousel',
            '.slide.bs.carousel',
            '.fade',
            '.fade.in',
            '.collapse',
            '.collapse.in',
            '.collapsing',
            '.alert-danger',
            '.logged-in .navbar-default',
            /^\.carousel-inner.*/,
            '#float-toc',
            '#float-toc a',
            /^\.modal-.*/,
            '.modal.fade.in',
            /(#|\.)modal(\-[a-zA-Z]+)?/,
            '.navbar-toggle.open',
            '.fade .modal-dialog',
            '.logged-in .navbar-fixed-top',
            '.navbar-fixed-top',
            '/^\.navbar-collapse.*/',
            '.navbar-inverse .innovations.navbar-toggle.open',
            '.single-innovation .navbar-inverse .innovations.navbar-toggle.open',
            '#innovations.collapse.in',
            'ul.page-numbers li a.prev',
            '.open',
            '.open > .dropdown-menu',
            '.open > a',
            '.alert-danger',
            '.visible-xs',
            '.noscript-warning',
            '.close',
            '.alert-dismissible',
            '.page.calendar .events .panel:hover .fa-angle-down.open',
            '.fa-angle-down.open',
            // needed for instafeed
            '.main #userfeed',
            '.main #userfeed a',
            '.main #userfeed:hover a',
            '.main #userfeed img'
          ],
          stylesheets: ['assets/css/main.css'],
          htmlroot: './_site/',
          media: ['(min-width: 992px)', '(min-width: 768px)', '(max-width: 767px)']
        },
        files: {
          'assets/css/main.css': ['_site/*.html', '_site/*/*.html', '_site/*/*/*.html']
        }
      }
    },
    shell: {
      jekyll: {
        command: 'bundle-2.5 exec jekyll build',
        options: {
          callback: log
        }
      },
      jekyll_incremental: {
        command: 'bundle-2.5 exec jekyll build --incremental',
        options: {
          callback: log
        }
      },
      htmlproofer: {
        command: 'bundle-2.5 exec htmlproofer --assume-extension --timeframe 7d --url-ignore=http://dx.doi.org/10.1038/ncb2018,http://dx.doi.org/10.1038/nrg2904 --check-html ./_site/',
        options: {
          callback: log
        }
      },
      updatecss: {
        command: 'cp assets/css/main.min.css _site/assets/css',
        options: {
          callback: log
        }
      }
    },
    watch: {
      sass: {
        files: 'scss/*.scss',
        tasks: ['build-css', 'dist-css'],
        options: {
          spawn: false
        }
      },
      jekyll_collections: {
        files: [
          '_people/*.md',
          '_pubs/*.md'
        ],
        tasks: 'shell:jekyll',
        options: {
          spawn: false
        }
      },
      jekyll: {
        files: [
          '_config.yml',
          'index.md',
          '_layouts/*',
          '_posts/*',
          '_includes/*',
          '_pages/*',
          'assets/js/*',
          'assets/pub/*',
          'assets/vid/*'
          // Omitting assets/css on purpose to prevent an endless update.
        ],
        tasks: 'shell:jekyll_incremental',
        options: {
          livereload: true,
          spawn: false
        }
      }
    },
    concurrent: {
      options: {
        logConcurrentOutput: true
      },
      watch: {
        tasks: [
          'watch:sass',
          'watch:jekyll_collections',
          'watch:jekyll']
      }
    },
    connect: {
      base: {
        options: {
          base: '_site',
          port: 3000,
          livereload: true,
          open: true
        }
      },
      keepalive: {
        options: {
          port: 3000,
          livereload: true,
          keepalive: true,
          open: true
        }
      }
    }
  });

  grunt.registerTask('build-css', 'build css', function() {
    var sassDest = '<%=builddir%>/css/main.css';
    var files = {};
    files[sassDest] = 'scss/main.scss';
    grunt.config('sass.dist.files', files);
    grunt.config('sass.dist.options.style', 'nested');
    grunt.task.run([
      'sass:dist',
      'prefix-css:' + sassDest,
      'compress-css:' + sassDest + ':<%=builddir%>/css/main.min.css'
    ]);
  });

  grunt.registerTask('dist-css', 'uncss and minify and update css in _site', function() {
    grunt.task.run([
      'uncss:dist',
      'compress-css:<%=builddir%>/css/main.css:<%=builddir%>/css/main.min.css',
      'shell:updatecss'
    ]);
  });

  grunt.registerTask('prefix-css', 'autoprefix a generic css', function(fileSrc) {
    grunt.config('autoprefixer.dist.src', fileSrc);
    grunt.task.run('autoprefixer');
  });

  grunt.registerTask('compress-css', 'compress a generic css', function(fileSrc, fileDst) {
    var files = {};
    files[fileDst] = fileSrc;
    grunt.config('sass.dist.files', files);
    grunt.config('sass.dist.options.style', 'compressed');
    grunt.task.run(['sass:dist']);
  });

  grunt.registerTask('build', [
    'concat:js',
    'uglify:js',
    'concat:funjs',
    'uglify:funjs',
    'build-css',
    'shell:jekyll',
    'dist-css', // XXX: Requires the generated html output, must be after jekyll.
    'clean:assets'
  ]);

  grunt.registerTask('serve', [
    'build',
    'connect:base',
    'concurrent:watch',
  ]);

  grunt.registerTask('default', 'serve');
};

// convert -resize "1000>" -strip -interlace JPEG -sampling-factor 4:2:0 -quality 70% -colorspace RGB src.jpeg dst.jpeg
